File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math
def solution(fees, records):
    answer = []
    b_t, b_f, p_t,p_f = fees
    parking = {}
    result = {}
    last_time = 23*60+59
    
    for record in records:
        t, car, io = record.split()
        h,m = map(int,t.split(':'))
        time = h*60+m
        if io == 'IN':
            parking[car] = time
        else:
            if car in result:
                result[car] += time-parking[car]
            else:
                result[car] = time-parking[car]
            del parking[car]
        
    for car, time in parking.items():
        if car in result:
            result[car] += last_time-parking[car]
        else:
            result[car] = last_time-parking[car]
            
    result = sorted(result.items(), key=lambda x:x[0])
   
    answer = []
    for car,time in result:
        answer.append(b_f + math.ceil(max(0,time-b_t)/p_t)*p_f)
    
    return answer

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
    var answer = [];
    let parking = new Map();
    let data = {};
    const [b_t, b_f, p_t, p_f] = fees
    for(let record of records){
        const [time, car, io] = record.split(' ')
        const [h,m] = time.split(':').map(e=> Number(e))
        t = h*60 + m
        
        if(io === 'IN'){
            parking.set(car,t)
        }
        else{
            if(data[car]){
                data[car]+=t - parking.get(car)
            }else{
                data[car] = t - parking.get(car)
            }
            parking.delete(car);
        }
    }
    
    for(let [car,time] of parking){
        if(data[car]){
            data[car]+=(23*60 + 59) - time
        }else{
            data[car]=(23*60 + 59) - time
        }
    }
    const sortedResult = Object.entries(data).sort((a, b) => a[0].localeCompare(b[0]));


    // 요금 계산
    for (const [car, time] of sortedResult) {
        const totalFee = b_f + Math.ceil(Math.max(0, time - b_t) / p_t) * p_f;
        answer.push(totalFee);
    }

    return answer;
}